scripts/lib/puppet_cache.py

```diff
@@ -0,0 +1,73 @@
+import hashlib
+import os
+import shutil
+import subprocess
+from typing import Optional
+
+import yaml
+
+from .zulip_tools import parse_os_release, run
+
+ZULIP_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+ZULIP_SRV_PATH = "/srv"
+
+PUPPET_MODULES_CACHE_PATH = os.path.join(ZULIP_SRV_PATH, "zulip-puppet-cache")
+PUPPET_DEPS_FILE_PATH = os.path.join(ZULIP_PATH, "puppet/deps.yaml")
+PUPPET_THIRDPARTY = os.path.join(PUPPET_MODULES_CACHE_PATH, "current")
+
+
+def generate_sha1sum_node_modules(
+    setup_dir: Optional[str] = None,
+) -> str:
+    sha1sum = hashlib.sha1()
+    with open(PUPPET_DEPS_FILE_PATH, "rb") as fb:
+        sha1sum.update(fb.read().strip())
+    sha1sum.update(subprocess.check_output(["puppet", "--version"]).strip())
+    return sha1sum.hexdigest()
+
+
+def setup_puppet_modules() -> None:
+    sha1sum = generate_sha1sum_node_modules()
+    target_path = os.path.join(PUPPET_MODULES_CACHE_PATH, sha1sum)
+    success_stamp = os.path.join(target_path, ".success-stamp")
+    # Check if a cached version already exists
+    if not os.path.exists(success_stamp):
+        do_puppet_module_install(target_path, success_stamp)
+
+    print("Using cached puppet modules from {}".format(target_path))
+    if os.path.islink(PUPPET_THIRDPARTY):
+        os.remove(PUPPET_THIRDPARTY)
+    elif os.path.isdir(PUPPET_THIRDPARTY):
+        shutil.rmtree(PUPPET_THIRDPARTY)
+    os.symlink(target_path, PUPPET_THIRDPARTY)
```

Comment:
What's the noop runtime of this logic?  That's worth optimizing, since we end up doing that wait every time we run puppet.  (We may find that removing the `yaml` import is a good idea, though I guess we already do that in puppet)

Reply:
Noop time is 2.3s for this function, entirely because of the `puppet --version` call.  If we're willing to assume that minor versions of puppet won't change the package resolution (which seems likely) then we can remove this entirely, and the noop takes 0.0004s.

If we want a faster version, we can switch to `ruby -r puppet/version -e 'puts Puppet.version'` which appears to be supported as an API, per `/usr/lib/ruby/vendor_ruby/puppet/version.rb`:
```
  # The intent is that software external to Puppet be able to determine the
  # Puppet version with no side-effects.  The expected use is:
  #
  #     require 'puppet/version'
  #     version = Puppet.version
```

Reply:
Yikes, even for `provision` / `upgrade` scripts, I would not want a 2.3s check.  140ms is also bad, but we can live with that.